openedx/core/djangoapps/bookmarks/tests/test_models.py

```diff
@@ -0,0 +1,432 @@
+            }),
+        )
+        self.bookmark_2 = BookmarkFactory.create(
+            user=self.user,
+            course_key=self.course_id,
+            usage_key=self.sequential_2.location,
+            xblock_cache=XBlockCache.create({
+                'display_name': self.sequential_2.display_name,
+                'usage_key': self.sequential_2.location,
+            }),
+        )
+
+        self.other_course = CourseFactory.create(display_name='An Introduction to API Testing 2')
+
+        self.other_chapter_1 = ItemFactory.create(
+            parent_location=self.other_course.location, category='chapter', display_name='Other Week 1'
+        )
+        self.other_sequential_1 = ItemFactory.create(
+            parent_location=self.other_chapter_1.location, category='sequential', display_name='Other Lesson 1'
+        )
+        self.other_sequential_2 = ItemFactory.create(
+            parent_location=self.other_chapter_1.location, category='sequential', display_name='Other Lesson 2'
+        )
+        self.other_vertical_1 = ItemFactory.create(
+            parent_location=self.other_sequential_1.location, category='vertical', display_name='Other Subsection 1'
+        )
+        self.other_vertical_2 = ItemFactory.create(
+            parent_location=self.other_sequential_1.location, category='vertical', display_name='Other Subsection 2'
+        )
+
+        # self.other_vertical_1 has two parents
+        self.other_sequential_2.children.append(self.other_vertical_1.location)
+        modulestore().update_item(self.other_sequential_2, self.admin.id)  # pylint: disable=no-member
+
+        self.other_bookmark_1 = BookmarkFactory.create(
+            user=self.user,
+            course_key=unicode(self.other_course.id),
+            usage_key=self.other_vertical_1.location,
+            xblock_cache=XBlockCache.create({
+                'display_name': self.other_vertical_1.display_name,
+                'usage_key': self.other_vertical_1.location,
+            }),
+        )
+
+    def create_course_with_blocks(self, children_per_block=1, depth=1, store_type=ModuleStoreEnum.Type.mongo):
+        """
+        Create a course and add blocks.
+        """
+        with self.store.default_store(store_type):
+
+            course = CourseFactory.create()
+            display_name = 0
+
+            with self.store.bulk_operations(course.id):
+                blocks_at_next_level = [course]
+
+                for level in range(depth):
+                    blocks_at_current_level = blocks_at_next_level
+                    blocks_at_next_level = []
+
+                    for block in blocks_at_current_level:
+                        for __ in range(children_per_block):
+                            blocks_at_next_level += [ItemFactory.create(
+                                parent_location=block.scope_ids.usage_id, display_name=unicode(display_name)
+                            )]
```

Comment:
@cpennington However, I am running into an issue here. When using the split store, multiple calls in the `for` loop to `ItemFactory.create()` return blocks which have the same `usage_id`. Any idea what could be causing this? This creates an incorrect tree.